File: src/direct/encoder.js
//@ts-check

import {
  FALSE,
  TRUE,

  UNDEFINED,
  NULL,

  NUMBER,
  NAN,
  INFINITY,
  N_INFINITY,
  ZERO,
  N_ZERO,

  BIGINT,
  STRING,
  SYMBOL,

  ARRAY,
  BUFFER,
  DATE,
  ERROR,
  MAP,
  OBJECT,
  REGEXP,
  SET,
  VIEW,

  RECURSION
} from './types.js';

import { toSymbol } from '../utils/symbol.js';

import { isArray, isView, push } from '../utils/index.js';
import { toTag } from '../utils/global.js';
import { encoder as textEncoder } from '../utils/text.js';
import { dv, u8a8, u8a4 } from './views.js';

/** @typedef {Map<number, number[]>} Cache */

const { isNaN, isFinite } = Number;
const { ownKeys } = Reflect;
const { is } = Object;

/**
 * @param {any} input
 * @param {number[]} output
 * @param {Cache} cache
 * @returns {boolean}
 */
const process = (input, output, cache) => {
  const value = cache.get(input);
  const unknown = !value;
  if (unknown) {
    dv.setUint32(0, output.length, true);
    cache.set(input, [...u8a4]);
  }
  else
    output.push(RECURSION, ...value);
  return unknown;
};

/**
 * @param {number[]} output
 * @param {number} type
 * @param {number} length
 */
const set = (output, type, length) => {
  dv.setUint32(0, length, true);
  output.push(type, ...u8a4);
};

/**
 * @param {any} input
 * @param {number[]} output
 * @param {Cache} cache
 */
const inflate = (input, output, cache) => {
  switch (typeof input) {
    case 'object': {
      if (input === null) {
        output.push(NULL);
        break;
      }
      else if (!process(input, output, cache)) break;
      else if (isArray(input)) {
        const length = input.length;
        set(output, ARRAY, length);
        for (let i = 0; i < length; i++)
          inflate(input[i], output, cache);
      }
      else if (isView(input)) {
        output.push(VIEW);
        inflate(toTag(input), output, cache);
        inflate(input.buffer, output, cache);
      }
      else if (input instanceof ArrayBuffer) {
        const ui8a = new Uint8Array(input);
        set(output, BUFFER, ui8a.length);
        push(output, ui8a);
      }
      else if (input instanceof Date) {
        output.push(DATE);
        inflate(input.getTime(), output, cache);
      }
      else if (input instanceof Map) {
        set(output, MAP, input.size);
        for (const [key, value] of input) {
          inflate(key, output, cache);
          inflate(value, output, cache);
        }
      }
      else if (input instanceof Set) {
        set(output, SET, input.size);
        for (const value of input)
          inflate(value, output, cache);
      }
      else if (input instanceof Error) {
        output.push(ERROR);
        inflate(input.name, output, cache);
        inflate(input.message, output, cache);
        inflate(input.stack, output, cache);
      }
      else if (input instanceof RegExp) {
        output.push(REGEXP);
        inflate(input.source, output, cache);
        inflate(input.flags, output, cache);
      }
      else {
        if ('toJSON' in input) {
          const json = input.toJSON();
          inflate(json === input ? null : json, output, cache);
        }
        else {
          const keys = ownKeys(input);
          const length = keys.length;
          set(output, OBJECT, length);
          for (let i = 0; i < length; i++) {
            const key = keys[i];
            inflate(key, output, cache);
            inflate(input[key], output, cache);
          }
        }
      }
      break;
    }
    case 'string': {
      if (process(input, output, cache)) {
        const encoded = textEncoder.encode(input);
        set(output, STRING, encoded.length);
        push(output, encoded);
      }
      break;
    }
    case 'symbol': {
      output.push(SYMBOL);
      inflate(toSymbol(input), output, cache);
      break;
    }
    case 'number':
      if (input && isFinite(input)) {
        dv.setFloat64(0, input, true);
        output.push(NUMBER, ...u8a8);
      }
      else if (isNaN(input)) output.push(NAN);
      else if (!input) output.push(is(input, 0) ? ZERO : N_ZERO);
      else output.push(input < 0 ? N_INFINITY : INFINITY);
      break;
    case 'boolean':
      output.push(input ? TRUE : FALSE);
      break;
    case 'bigint': {
      dv.setBigInt64(0, input, true);
      output.push(BIGINT, ...u8a8);
      break;
    }
    // this covers functions too
    default:
      output.push(UNDEFINED);
      break;
  }
};

/**
 * @param {any} value
 * @returns {number[]}
 */
export const encode = value => {
  const output = [];
  inflate(value, output,  new Map);
  return output;
};

/**
 * @param {{ byteOffset?: number }} [options]
 * @returns {(value: any, buffer: SharedArrayBuffer) => number}
 */
export const encoder = ({ byteOffset = 0 } = {}) => (value, buffer) => {
  const output = encode(value);
  const length = output.length;
  const size = length + byteOffset;
  if (buffer.byteLength < size) {
    //@ts-ignore
    buffer.grow(size);
  }
  new Uint8Array(buffer, byteOffset, length).set(output);
  return length;
};

File: src/direct/encoder.py
import struct
from datetime import datetime, timezone

from .js import Blob, File, Map, Null, Set, Symbol
from .types import FALSE, TRUE, NULL, UNDEFINED, NUMBER, UI8, NAN, INFINITY, N_INFINITY, ZERO, N_ZERO, BIGINT, BIGUINT, STRING, SYMBOL, ARRAY, BUFFER, DATE, ERROR, MAP, OBJECT, REGEXP, SET, VIEW, IMAGE_DATA, BLOB, FILE, FOREIGN_ARRAY, FOREIGN_SET, RECURSION
from .views import dv, u8a8

inf = float('inf')
ninf = float('-inf')
nan = float('nan')

class Cache:
  def __init__(self):
    self.i = []
    self.v = []

def append(output, type, length):
  dv.setUint32(0, length, True)
  output.append(type)
  output.extend(u8a8[0:4])

def append_array(type, input, output, cache):
  append(output, type, len(input))
  for value in input:
    inflate(value, output, cache)

def append_set(type, input, output, cache):
  append(output, type, len(input))
  for value in input:
    inflate(value, output, cache)

def append_object(type, input, output, cache):
  items = [[k, v] for k, v in input.items()]
  append(output, type, len(items))
  for key, value in items:
    inflate(key, output, cache)
    inflate(value, output, cache)

def process(input, output, cache):
  try:
    index = cache.i.index(input)
    output.append(RECURSION)
    output.extend(cache.v[index])
    return False
  except:
    dv.setUint32(0, len(output), True)
    cache.i.append(input)
    cache.v.append(u8a8[0:4])
    return True

def inflate(input, output, cache):
  if input is None:
    output.append(UNDEFINED)

  elif input is Null:
    output.append(NULL)

  elif input is True:
    output.append(TRUE)

  elif input is False:
    output.append(FALSE)

  elif isinstance(input, int):
    if input < 256 and -1 < input:
      output.append(UI8)
      output.append(input)
    else:
      if 9223372036854775807 < input:
        output.append(BIGUINT)
        dv.setBigUint64(0, input, True)
      elif input < -9007199254740991 or input > 9007199254740991:
        output.append(BIGINT)
        dv.setBigInt64(0, input, True)
      else:
        output.append(NUMBER)
        dv.setFloat64(0, input, True)
      output.extend(u8a8)

  elif isinstance(input, float):
    if input == inf:
      output.append(INFINITY)
    elif input == ninf:
      output.append(N_INFINITY)
    elif input == nan:
      output.append(NAN)
    else:
      if not input:
        output.append(ZERO if struct.pack('f', input)[3] == 0 else N_ZERO)
      else:
        output.append(NUMBER)
        dv.setFloat64(0, input, True)
        output.extend(u8a8)

  elif isinstance(input, Symbol):
    output.append(SYMBOL)
    inflate(str(input), output, cache)

  elif process(input, output, cache):
    if isinstance(input, str):
      utf8 = input.encode('utf-8')
      append(output, STRING, len(utf8))
      output.extend(utf8)

    elif isinstance(input, Set):
      append_set(SET, input, output, cache)

    elif isinstance(input, Map):
      append_object(MAP, input, output, cache)

    elif isinstance(input, set):
      append_set(FOREIGN_SET, input, output, cache)

    elif isinstance(input, tuple):
      append_array(FOREIGN_ARRAY, input, output, cache)

    elif isinstance(input, list):
      append_array(ARRAY, input, output, cache)

    elif isinstance(input, dict):
      append_object(OBJECT, input, output, cache)

    elif isinstance(input, memoryview):
      output.append(VIEW)
      inflate('Uint8Array', output, cache)
      # TODO: shared bytearray here causes issues in cache
      append(output, BUFFER, len(input))
      output.extend(input)

    elif isinstance(input, (bytes, bytearray)):
      append(output, BUFFER, len(input))
      output.extend(input)

def encode(value):
  output = bytearray()
  inflate(value, output, Cache())
  return output

def encoder(byteOffset = 0):
  def _(value):
    output = encode(value)
    length = len(output)
    if byteOffset > 0:
      output = ([0] * byteOffset) + output
    return [length, output]

  return _
